pubsub/stanpubsub/stan_test.go

```diff
@@ -0,0 +1,609 @@
+		t.Fatalf("Expected %v, got %v", gcerrors.FailedPrecondition, gce)
+	}
+	if gce := qs.ErrorCode(nats.ErrAuthorization); gce != gcerrors.PermissionDenied {
+		t.Fatalf("Expected %v, got %v", gcerrors.PermissionDenied, gce)
+	}
+	if gce := qs.ErrorCode(nats.ErrMaxMessages); gce != gcerrors.ResourceExhausted {
+		t.Fatalf("Expected %v, got %v", gcerrors.ResourceExhausted, gce)
+	}
+	if gce := qs.ErrorCode(nats.ErrSlowConsumer); gce != gcerrors.ResourceExhausted {
+		t.Fatalf("Expected %v, got %v", gcerrors.ResourceExhausted, gce)
+	}
+	if gce := qs.ErrorCode(nats.ErrTimeout); gce != gcerrors.DeadlineExceeded {
+		t.Fatalf("Expected %v, got %v", gcerrors.DeadlineExceeded, gce)
+	}
+}
+
+func BenchmarkNatsStreamingQueuePubSub(b *testing.B) {
+	ctx := context.Background()
+
+	o := nsserver.GetDefaultOptions()
+	opts := nsserver.DefaultNatsServerOptions
+	opts.Port = benchPort
+	o.ID = testClusterID
+	s, err := nsserver.Run(o, &opts)
+	if err != nil {
+		b.Fatalf("nats-streaming: %v", err)
+	}
+	defer s.Shutdown()
+
+	nc, err := stan.Connect(testClusterID, testClientID, stan.NatsURL(fmt.Sprintf("nats://127.0.0.1:%d", benchPort)))
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer nc.Close()
+
+	h := &harness{ns: s, sc: nc, queueGroup: b.Name()}
+	dt, cleanup, err := h.CreateTopic(ctx, b.Name())
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer cleanup()
+
+	qs, cleanup, err := h.CreateSubscription(ctx, dt, b.Name())
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer cleanup()
+
+	recvBatcherOpts := &batcher.Options{MaxBatchSize: stan.DefaultMaxInflight}
+
+	topic := pubsub.NewTopic(dt, nil)
+	defer topic.Shutdown(ctx)
+	queueSub := pubsub.NewSubscription(qs, recvBatcherOpts, nil)
+	defer queueSub.Shutdown(ctx)
+
+	drivertest.RunBenchmarks(b, topic, queueSub)
+}
+
+func BenchmarkNatsStreamingPubSub(b *testing.B) {
+	ctx := context.Background()
+
+	o := nsserver.GetDefaultOptions()
+	opts := nsserver.NewNATSOptions()
+	opts.Port = benchPort
+	o.ID = testClusterID
+	s, err := nsserver.Run(o, opts)
+	if err != nil {
+		b.Fatalf("nats-streaming: %v", err)
+	}
+	defer s.Shutdown()
+
+	nc, err := stan.Connect(testClusterID, testClientID, stan.NatsURL(fmt.Sprintf("nats://127.0.0.1:%d", benchPort)))
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer nc.Close()
+
+	h := &harness{ns: s, sc: nc}
+	dt, cleanup, err := h.CreateTopic(ctx, b.Name())
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer cleanup()
+	ds, cleanup, err := h.CreateSubscription(ctx, dt, b.Name())
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer cleanup()
+
+	topic := pubsub.NewTopic(dt, nil)
+	defer topic.Shutdown(ctx)
+
+	recvBatcherOpts := &batcher.Options{MaxBatchSize: stan.DefaultMaxInflight}
+	sub := pubsub.NewSubscription(ds, recvBatcherOpts, nil)
+	defer sub.Shutdown(ctx)
+
+	drivertest.RunBenchmarks(b, topic, sub)
+}
+
+func testOpenerConnection() func() {
+	// NOTE: Due to the nature of the NATS Streaming - the reconnection prevents any buffering - which may cause a published
```

Comment:
To avoid this, can we instead make this function call `connect` and return an `*URLOpener, func()`? Then the callers can use that `URLOpener` instead of the default one.

Reply:
As I've mentioned in previous comments the `connect` function provided a really clean way of defining `URLOpener`. 
Thanks.